content/ember/v5/deprecate-array-prototype-extensions.md

```diff
@@ -0,0 +1,1265 @@
+  @tracked abc = ['x', 'y', 'z', 'x'];
+
+  @action
+  clear() {
+    this.abc = [];
+  }
+}
+```
+
+#### `insertAt`
+
+Before
+```js
+import Component from '@glimmer/component';
+export default class SampleComponent extends Component {
+  abc = ['x', 'y', 'z', 'x'];
+
+  @action
+  insertAt(idx, value) {
+    this.abc.insertAt(idx, value);
+  }
+}
+```
+
+After
+```js
+// TrackedArray
+import Component from '@glimmer/component';
+import { action } from '@ember/object';
+import { TrackedArray } from 'tracked-built-ins';
+
+export default class SampleComponent extends Component {
+  abc = new TrackedArray(['x', 'y', 'z', 'x']);
+
+  @action
+  insertAt(idx, value) {
+    this.abc.splice(idx, 0, value);
+  }
+}
+```
+or
+
+```js
+// @tracked
+import Component from '@glimmer/component';
+import { action } from '@ember/object';
+import { tracked } from '@glimmer/tracking';
+
+export default class SampleComponent extends Component {
+  @tracked abc = ['x', 'y', 'z', 'x'];
+
+  @action
+  insertAt(idx, value) {
+    this.abc = [...this.abc.slice(0, idx), value, this.abc.slice(this.abc.length - idx)]
+  }
+}
+```
+
+#### `popObject`
+Before
+```js
+import Component from '@glimmer/component';
+export default class SampleComponent extends Component {
+  abc = ['x', 'y', 'z', 'x'];
+
+  @action
+  popObject() {
+    this.abc.popObject();
+  }
+}
+```
+
+After
+```js
+// TrackedArray
+import Component from '@glimmer/component';
+import { action } from '@ember/object';
+import { TrackedArray } from 'tracked-built-ins';
+
+export default class SampleComponent extends Component {
+  abc = new TrackedArray(['x', 'y', 'z', 'x']);
+
+  @action
+  popObject() {
+    this.abc.pop();
+  }
+}
+```
+or
+
+```js
+// @tracked
+import Component from '@glimmer/component';
+import { action } from '@ember/object';
+import { tracked } from '@glimmer/tracking';
+
+export default class SampleComponent extends Component {
+  @tracked abc = ['x', 'y', 'z', 'x'];
+
+  @action
+  popObject() {
+    this.abc.pop();
+    this.abc = [...this.abc];
+  }
+}
+```
+
+#### `pushObject`
+Before
+```js
+import Component from '@glimmer/component';
+export default class SampleComponent extends Component {
+  abc = ['x', 'y', 'z', 'x'];
+
+  @action
+  pushObject(value) {
+    this.abc.pushObject(value);
+  }
+}
+```
+
+After
+```js
+// TrackedArray
+import Component from '@glimmer/component';
+import { action } from '@ember/object';
+import { TrackedArray } from 'tracked-built-ins';
+
+export default class SampleComponent extends Component {
+  abc = new TrackedArray(['x', 'y', 'z', 'x']);
+
+  @action
+  pushObject(value) {
+    this.abc.push(value);
+  }
+}
+```
+or
+
+```js
+// @tracked
+import Component from '@glimmer/component';
+import { action } from '@ember/object';
+import { tracked } from '@glimmer/tracking';
+
+export default class SampleComponent extends Component {
+  @tracked abc = ['x', 'y', 'z', 'x'];
+
+  @action
+  pushObject(value) {
+    this.abc = [...this.abc, value];
```

Comment:
I'm a little confused by the difference between this example and the `pushObject` example above which uses `TrackedArray`. Is it a requirement to assign to `this.abc` when using `@tracked` but not when using `TrackedArray`? Or is the different method shown to indicate that both approaches are valid?

Reply:
yeah, `@tracked` will only trigger the reactivity when a new value gets assigned. When the existing native array is internally mutated, like calling `.push()`, then this will not be reactive. In opposite to `TrackedArray`, which does this kind of deep tracking. So I think the example is right.